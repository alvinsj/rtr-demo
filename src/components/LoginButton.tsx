import { useCallback, useEffect, useState } from 'react'
import { useContextSelector } from 'use-context-selector'

import useInitPKCE from '@/hooks/useInitPKCE'
import AuthContext from '@/contexts/AuthContext'
import { deleteRefreshToken } from '@/utils/refreshToken'

type LoginButtonProps = {
  className?: string
}
const LoginButton = ({ className }: LoginButtonProps) => {
  const accessToken = useContextSelector(
    AuthContext, ctx => ctx.accessToken
  )
  const setAccessToken = useContextSelector(
    AuthContext, ctx => ctx.setAccessToken
  )
  const [isLoading, setIsLoading] = useState(false)

  const { error, onLogin } = useInitPKCE()
  const handleLogout = useCallback(() => {
    deleteRefreshToken()
    setAccessToken(null)
  }, [setAccessToken])

  useEffect(() => {
    const handleRouteChangeStart = () => setIsLoading(true)
    window.addEventListener('beforeunload', handleRouteChangeStart)
    return () => window.removeEventListener('beforeunload', handleRouteChangeStart)
  }, [])

  const isLoggedIn = !!accessToken
  const buttonText = isLoggedIn ? 'Logout' : 'Login'
  const handleButtonClick = isLoggedIn ? handleLogout : onLogin

  return (
    <div className={className}>
      <span className='error'>{error}</span>
      <button type="submit"
        onClick={handleButtonClick}
        disabled={isLoading}>
        {isLoading ? 'Loading...' : buttonText}
      </button>
    </div>
  )
}

export default LoginButton
