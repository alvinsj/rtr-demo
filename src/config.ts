const LOGIN_URL = import.meta.env.VITE_AUTH_URL
const TOKEN_URL = import.meta.env.VITE_TOKEN_URL
const BASE_URL = import.meta.env.VITE_BASE_URL

export default {
  LOGIN_URL,
  TOKEN_URL,
  BASE_URL
}
