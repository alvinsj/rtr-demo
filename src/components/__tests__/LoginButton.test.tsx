
import { test, expect, vi, describe, Mock } from 'vitest'
import { render, fireEvent, act } from '@testing-library/react'

import LoginButton from '../LoginButton'
import config from '@/config'

const mocks = vi.hoisted(() => {
  return {
    useContextSelector: vi.fn(),
    deleteRefreshToken: vi.fn(),
    setAccessToken: vi.fn()
  }
})
vi.mock('use-context-selector', async (importOriginal) => {
  const actual = await importOriginal()
  return ({
    ...(actual as Record<string, unknown>),
    useContextSelector: mocks.useContextSelector
  })
})
vi.mock('@/utils/refreshToken', async (importOriginal) => {
  const actual = await importOriginal()
  return ({
    ...(actual as Record<string, unknown>),
    deleteRefreshToken: mocks.deleteRefreshToken
  })
})

describe('LoginButton', () => {
  test('renders with snapshot', () => {
    mocks.useContextSelector.mockReturnValue(null)

    const screen = render(<LoginButton />)
    expect(screen.container.children[0]).toMatchSnapshot()
  })

  test('logs in and redirects to login url', async () => {
    mocks.useContextSelector.mockReturnValue(null)
    vi.spyOn(location, 'replace')

    const { getByText } = render(<LoginButton />)
    fireEvent.click(getByText(/Login/))

    await vi.waitFor(() => {
      expect(location.replace).toHaveBeenCalled()

      const urlString = (location.replace as Mock).mock.calls[0][0]
      const url = new URL(urlString)
      expect(url.origin + url.pathname).toBe(config.LOGIN_URL)
    })
  })

  test('logs out and clears tokens', async () => {
    mocks.useContextSelector.mockReturnValueOnce('access_token_jwt')
    mocks.useContextSelector.mockReturnValueOnce(mocks.setAccessToken)

    const { getByText } = render(<LoginButton />)
    expect(getByText(/Logout/))

    fireEvent.click(getByText(/Logout/))
    expect(mocks.deleteRefreshToken).toHaveBeenCalledWith()
    expect(mocks.setAccessToken).toHaveBeenCalledWith(null)
  })

  test('shows loading state', async () => {
    mocks.useContextSelector.mockReturnValue(null)

    const { getByText } = render(<LoginButton />)
    expect(getByText(/Login/))

    act(() => {
      const event = new Event('beforeunload', {
        bubbles: true,
        cancelable: true,
      })
      window.dispatchEvent(event)
    })

    expect(getByText(/Loading/))
  })
})




