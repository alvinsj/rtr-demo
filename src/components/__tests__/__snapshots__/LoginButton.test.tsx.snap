// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginButton > renders with snapshot 1`] = `
<div>
  <span
    class="error"
  />
  <button
    type="submit"
  >
    Login
  </button>
</div>
`;
